generate/lora.py

```diff
@@ -0,0 +1,154 @@
+    quantize: Literal["llm.int8", "gptq.int4"] = None,
+    max_new_tokens: int = 100,
+    top_k: int = 200,
+    temperature: float = 0.8,
+    strategy: str = "auto",
+    devices: int = 1,
+    precision: str = "bf16-true",
+) -> None:
+    """Generates a response based on a given instruction and an optional input.
+    This script will only work with checkpoints from the instruction-tuned Parrot-Adapter model.
+    See `finetune/adapter.py`.
+
+    Args:
+        prompt: The prompt/instruction (Alpaca style).
+        adapter_path: Path to the checkpoint with trained adapter weights, which are the output of
+            `finetune/adapter.py`.
+        checkpoint_dir: The path to the checkpoint folder with pretrained Parrot weights.
+        input: Optional input (Alpaca style).
+        quantize: Whether to quantize the model and using which method:
+            ``"llm.int8"``: LLM.int8() mode,
+            ``"gptq.int4"``: GPTQ 4-bit mode.
+        max_new_tokens: The number of generation steps to take.
+        top_k: The number of top most probable tokens to consider in the sampling process.
+        temperature: A value controlling the randomness of the sampling process. Higher values result in more random
+            samples.
+        strategy: Indicates the Fabric strategy setting to use.
+        devices: How many devices to use.
+        precision: Indicates the Fabric precision setting to use.
+    """
+    if strategy == "fsdp":
+        auto_wrap_policy = partial(transformer_auto_wrap_policy, transformer_layer_cls={Block})
+        strategy = FSDPStrategy(auto_wrap_policy=auto_wrap_policy, cpu_offload=False)
+    fabric = L.Fabric(devices=devices, precision=precision, strategy=strategy)
+    fabric.launch()
+
+    check_valid_checkpoint_dir(checkpoint_dir)
+
+    with open(checkpoint_dir / "lit_config.json") as fp:
+        config = Config(**json.load(fp))
+
+    if quantize is not None:
+        raise NotImplementedError("Quantization in LoRA is not supported yet")
+    else:
+        model_file = "lit_model.pth"
+    # Remove the above, and then enable this later:
+    """
+    if quantize is not None and devices > 1:
+        raise NotImplementedError
+    if quantize == "gptq.int4":
+        model_file = "lit_model_gptq.4bit.pth"
+        if not (checkpoint_dir / model_file).is_file():
+            raise ValueError("Please run `python quantize/gptq.py` first")
+    else:
+        model_file = "lit_model.pth"
+    """
+    checkpoint_path = checkpoint_dir / model_file
+
+    fabric.print(f"Loading model {str(checkpoint_path)!r} with {config.__dict__}", file=sys.stderr)
+    t0 = time.time()
+    with fabric.init_module(), EmptyInitOnDevice(quantization_mode=quantize):
+        model = Parrot(config)
+    fabric.print(f"Time to instantiate model: {time.time() - t0:.02f} seconds.", file=sys.stderr)
+
+    t0 = time.time()
+
+    with lazy_load(checkpoint_path) as pretrained_checkpoint, lazy_load(lora_path) as lora_checkpoint:
```

Comment:
Lazy loading should be done after the model is loaded, as in the other generation scripts